runtime/instrtrace.c

```diff
@@ -23,27 +23,34 @@
 #include <string.h>
 #include <ctype.h>
 
+#include "caml/fiber.h"
+#include "caml/domain.h"
 #include "caml/instrtrace.h"
 #include "caml/instruct.h"
 #include "caml/misc.h"
 #include "caml/mlvalues.h"
 #include "caml/opnames.h"
 #include "caml/prims.h"
-#include "caml/stacks.h"
-#include "caml/startup_aux.h"
+#include "caml/startup.h"
 
 extern code_t caml_start_code;
 
-intnat caml_icount = 0;
+__thread intnat caml_icount = 0;
 
-void caml_stop_here () {}
+void caml_stop_here (void)
+{
+}
 
-void caml_disasm_instr(pc)
-     code_t pc;
+char * caml_instr_string (code_t pc);
+
+void caml_disasm_instr(code_t pc)
 {
+  char buf[256];
+  char opbuf[128];
   int instr = *pc;
-  printf("%6ld  %s", (long) (pc - caml_start_code),
-         instr < 0 || instr > STOP ? "???" : names_of_instructions[instr]);
+  snprintf(opbuf, sizeof(opbuf), "%6ld  %s", (long) (pc - caml_start_code),
+           (instr < 0 || instr >= FIRST_UNIMPLEMENTED_OP) ?
```

Comment:
Trunk uses `STOP` in the comparison whereas multicore uses `FIRST_UNIMPLEMENTED_OP`. Is there a reason for to have this difference?

Reply:
`FIRST_UNIMPLEMENTED_OP` is better.

Reply:
To elaborate: a long time ago, `STOP` was the instruction with the highest opcode, but this forced a renumbering of instructions when new instructions were added, causing a bit of extra complexity in the bootstrap.  So, now we can add instructions "after" `STOP`, and `FIRST_UNIMPLEMENTED_OP` keeps track of this.

Reply:
Thanks for the clarification.